Use WeakPtr in WeakHashSet

geoffreygaren opened the pull request:
#### f8a1a8119a8ecf6eec0d46260f5f0c8a6603e0f7
<pre>
Use WeakPtr in WeakHashSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=303247">https://bugs.webkit.org/show_bug.cgi?id=303247</a>
&lt;<a href="https://rdar.apple.com/problem/165538736">rdar://problem/165538736</a>&gt;

Reviewed by Darin Adler.

This removes the need for WeakHashSet to be a friend to WeakPtr
and WeakPtrFactory.

This is also a step toward a more efficient WeakPtr.

I added a null key check to the HashTable consistency check because a WeakPtr&lt;T&gt;
key can become null during normal operation. It&apos;s impossible for a valid key
lookup to find the null, but manual bucket iteration, as in the HashTable
consistency check, will find it.

I amended WeakPtr hash traits to make sure not to instantiate a WeakPtr when
checking for the empty and deleted values. This is a small optimization and
it&apos;s required to pass API Tests.

* Source/WTF/wtf/HashTable.h:
(WTF::Malloc&gt;::checkTableConsistencyExceptSize const):
* Source/WTF/wtf/WeakHashSet.h:
(WTF::containerSize):
(WTF::copyToVector):
* Source/WTF/wtf/WeakPtr.h:
(WTF::WeakPtrHashTraits::isEmptyValue):
(WTF::WeakPtrHashTraits::isDeletedValue):
* Source/WTF/wtf/WeakPtrFactory.h:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::isElementWithPendingSVGResources const):

Canonical link: <a href="https://commits.webkit.org/303699@main">https://commits.webkit.org/303699@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b38e3a43353f07546bfbeee545af6ed8a4957c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133402 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5903 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44539 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140958 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/79fb0828-2bcc-43ac-bdca-215a155222c6) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135272 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6416 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5768 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102051 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/585b317b-33d1-4f5d-b17f-c4137565055e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136349 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/119582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82845 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bcfeca2e-7ee7-4a2b-aff2-ee16a7c19c99) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125474 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/113533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/37692 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143604 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/131913 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5573 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38279 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/5655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4783 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110608 "Passed tests") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115841 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20629 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5628 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164878 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/5474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69080 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43080 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/5717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/5584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/HashTable.h:
```diff
@@ -1442,6 +1442,11 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(HashTable);
             }
 
             auto& key = Extractor::extract(*entry);
+            if (!key) {
+                ++count;
+                continue;
+            }
```
Does this need a comment, or is it obvious why it’s needed?

geoffreygaren replied:
I'll add this:

`            // A weak key can become null without being eagerly removed from the table.`

darinadler commented on Source/WTF/wtf/WeakHashSet.h:
```diff
@@ -196,17 +195,17 @@ class WeakHashSet final {
 
     void forEach(NOESCAPE const Function<void(T&)>& callback) requires WTF::HasRefPtrMemberFunctions<T>::value
```
Surprised to see `WTF::` prefixes here inside the `WTF` namespace.

geoffreygaren replied:
Will fix.

geoffreygaren committed:
Use WeakPtr in WeakHashSet
https://bugs.webkit.org/show_bug.cgi?id=303247
<rdar://problem/165538736>

Reviewed by Darin Adler.

This removes the need for WeakHashSet to be a friend to WeakPtr
and WeakPtrFactory.

This is also a step toward a more efficient WeakPtr.

I added a null key check to the HashTable consistency check because a WeakPtr<T>
key can become null during normal operation. It's impossible for a valid key
lookup to find the null, but manual bucket iteration, as in the HashTable
consistency check, will find it.

I amended WeakPtr hash traits to make sure not to instantiate a WeakPtr when
checking for the empty and deleted values. This is a small optimization and
it's required to pass API Tests.

* Source/WTF/wtf/HashTable.h:
(WTF::Malloc>::checkTableConsistencyExceptSize const):
* Source/WTF/wtf/WeakHashSet.h:
(WTF::containerSize):
(WTF::copyToVector):
* Source/WTF/wtf/WeakPtr.h:
(WTF::WeakPtrHashTraits::isEmptyValue):
(WTF::WeakPtrHashTraits::isDeletedValue):
* Source/WTF/wtf/WeakPtrFactory.h:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::isElementWithPendingSVGResources const):

Canonical link: https://commits.webkit.org/303699@main
